### ISSUE TITLE
Block components need to know when the component is focused / selected

### ISSUE THREAD

**Issue by DavidC-Mega** [WordPress/gutenberg#42907]
## The problem

There is no way to show controls when individual custom block components are selected.
I have created some custom block components that extend `wp.element.Component`.
I would like to show extra controls (e.g. via toolbar or inspector) when these custom components are selected. I see this is already occurring with the `RichText` component:

![richtext-instance-settings](https://user-images.githubusercontent.com/86385557/182529012-4e396cbc-7988-4389-bca6-8c82c1019a75.png)
In this screenshot you can see a custom block with two RichText components: the RichText toolbar controls are based on the selected RichText component (for the header, I have no `allowedFormats`) and a custom button component.
I would like to be able to show settings when this button component is selected in similar fashion to the `RichText` elements.
At the moment, all controls for all custom components within the blocks are displayed.

## Proposed solution

Expose a way for custom block components to be selected/focused.
Some more documentation on custom components for use in blocks would also be greatly appreciated.

**Comment by t-hamano** [comment]
As for the `RichText` component, `unstableOnFocus` can catch the focus being applied (it is non-stable).
Although it may not be ideal, the icons on the block toolbar can be switched depending on the focus state, as shown below.
It might be possible to control your own custom components as well `onFocus`.

```javascript
import {
	useBlockProps,
	BlockControls,
	RichText,
} from '@wordpress/block-editor';
import { useState } from '@wordpress/element';
import { ToolbarButton } from '@wordpress/components';

export default function Edit() {
	const [ acitveComponent, setActiveComponent ] = useState();

	return (
		<>
			<BlockControls>
				{ acitveComponent === 'paragraph' && (
					<ToolbarButton label="Paragraph" icon="editor-paragraph" />
				) }
				{ acitveComponent === 'button' && (
					<ToolbarButton label="Button" icon="button" />
				) }
			</BlockControls>
			<div { ...useBlockProps() }>
				<RichText
					unstableOnFocus={ () => setActiveComponent( 'paragraph' ) }
				/>
				<RichText
					unstableOnFocus={ () => setActiveComponent( 'button' ) }
				/>
			</div>
		</>
	);
}
```

**Comment by DavidC-Mega** [comment]
Thanks @t-hamano for your comment.

I managed to get what I wanted by using state ("isSelected") in each custom component and updating the state by onFocus / onBlur listeners attached to a wrapper div component. I'm sure I had tried this earlier, but perhaps not correctly :thinking:

Anyway, works fine! Inspector options are now hidden for components within the block not selected.

So I think we can close this, thanks :slightly_smiling_face:
